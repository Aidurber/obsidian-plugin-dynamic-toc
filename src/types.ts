export type BulletStyle = "bullet" | "number";
export interface TableOptions {
  style: BulletStyle;
  min_depth: number;
  max_depth: number;
}

export interface DynamicTOCSettings extends TableOptions {}
