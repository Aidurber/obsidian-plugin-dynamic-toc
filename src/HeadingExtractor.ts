import { App, TFile } from "obsidian";
import { DynamicTOCSettings, TableOptions } from "./types";

export interface IHeadingExtractor {
  extract(file: TFile, options: TableOptions): string;
}
export class HeadingExtractor implements IHeadingExtractor {
  constructor(protected app: App) {}
  extract(file: TFile, options: TableOptions): string {
    const { headings } = this.app.metadataCache.getFileCache(file);
    const processableHeadings = headings.filter(
      (h) => h.level >= options.min_depth && h.level <= options.max_depth
    );
    const firstHeadingDepth = processableHeadings[0].level;
    return processableHeadings
      .map((heading) => {
        const itemIndication = (options.style === "number" && "1.") || "-";
        const indent = new Array(Math.max(0, heading.level - firstHeadingDepth))
          .fill("\t")
          .join("");
        return `${indent}${itemIndication} [[#${heading.heading}|${heading.heading}]]`;
      })
      .join("\n");
  }
}
