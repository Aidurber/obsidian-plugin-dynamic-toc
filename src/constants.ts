import {
  DynamicTOCSettings,
  ExternalMarkdownKey,
  EXTERNAL_MARKDOWN_PREVIEW_STYLE,
} from "./types";

export const DEFAULT_SETTINGS: DynamicTOCSettings = {
  style: "bullet",
  min_depth: 2,
  max_depth: 6,
  externalStyle: "None",
  supportAllMatchers: false,
};

export const TABLE_CLASS_NAME = "dynamic-toc";
export const TABLE_CLASS_SELECTOR = `.${TABLE_CLASS_NAME}`;

export const ALL_MATCHERS = Object.keys(
  EXTERNAL_MARKDOWN_PREVIEW_STYLE
) as ExternalMarkdownKey[];
