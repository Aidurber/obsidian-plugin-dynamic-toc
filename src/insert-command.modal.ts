import { App, FuzzySuggestModal } from "obsidian";
import DynamicTOCPlugin from "./main";
import { ExternalMarkdownKey } from "./types";

type OptionsCollection = Record<
  Exclude<ExternalMarkdownKey, "None"> & "code-block",
  {
    label: string;
    value: string;
  }
>;

// TODO refactor to use this as external matchers value so we have a single source of truth
const options: OptionsCollection = {
  "code-block": { value: `\`\`\`toc\n\`\`\``, label: "Code block" },
  TOC: { value: "[TOC]", label: "[TOC]" },
  _TOC_: { label: "__TOC__", value: "[[__TOC__]]" },
  AzureWiki: { label: "_TOC_", value: "[[_TOC_]]" },
  DevonThink: { label: "{{toc}}", value: "{{toc}}" },
  TheBrain: { label: "[/toc/]", value: "[/toc/]" },
};
export class InsertCommandModal extends FuzzySuggestModal<string> {
  private plugin: DynamicTOCPlugin;
  callback: (item: string) => void;
  constructor(app: App, plugin: DynamicTOCPlugin) {
    super(app);
    this.app = app;
    this.plugin = plugin;
    this.setPlaceholder("Type name of table of contents type...");
  }
  getItems(): string[] {
    if (this.plugin.settings.supportAllMatchers) {
      return Object.keys(options);
    }
    if (this.plugin.settings.externalStyle !== "None") {
      return ["code-block", this.plugin.settings.externalStyle];
    }
    return ["code-block"];
  }
  getItemText(id: string): string {
    const foundKey = Object.keys(options).find((v) => v === id);
    return options[foundKey].label;
  }
  onChooseItem(item: string): void {
    this.callback(options[item].value);
  }
  public start(callback: (item: string) => void): void {
    this.callback = callback;
    this.open();
  }
}
