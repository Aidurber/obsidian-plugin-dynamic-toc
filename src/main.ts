import {
  App,
  Modal,
  Notice,
  Plugin,
  PluginSettingTab,
  Setting,
} from "obsidian";

interface MyPluginSettings {
  mySetting: string;
}

const DEFAULT_SETTINGS: MyPluginSettings = {
  mySetting: "default",
};

export default class MyPlugin extends Plugin {
  settings: MyPluginSettings;

  async onload() {
    console.log("loading plugin");

    await this.loadSettings();

    this.addRibbonIcon("dice", "Sample Plugin", () => {
      new Notice("This is a notice!");
    });

    this.addStatusBarItem().setText("Status Bar Text");

    this.addCommand({
      id: "open-sample-modal",
      name: "Open Sample Modal",
      // callback: () => {
      // 	console.log('Simple Callback');
      // },
      checkCallback: (checking: boolean) => {
        let leaf = this.app.workspace.activeLeaf;
        if (leaf) {
          if (!checking) {
            new SampleModal(this.app).open();
          }
          return true;
        }
        return false;
      },
    });

    this.addSettingTab(new SampleSettingTab(this.app, this));

    this.registerCodeMirror((cm: CodeMirror.Editor) => {
      console.log("codemirror", cm);
    });

    this.registerDomEvent(document, "click", (evt: MouseEvent) => {
      console.log("click", evt);
    });

    this.registerInterval(
      window.setInterval(() => console.log("setInterval"), 5 * 60 * 1000)
    );
  }

  onunload() {
    console.log("unloading plugin");
  }

  async loadSettings() {
    this.settings = Object.assign({}, DEFAULT_SETTINGS, await this.loadData());
  }

  async saveSettings() {
    await this.saveData(this.settings);
  }
}

class SampleModal extends Modal {
  constructor(app: App) {
    super(app);
  }

  onOpen() {
    let { contentEl } = this;
    contentEl.setText("Woah!");
  }

  onClose() {
    let { contentEl } = this;
    contentEl.empty();
  }
}

class SampleSettingTab extends PluginSettingTab {
  plugin: MyPlugin;

  constructor(app: App, plugin: MyPlugin) {
    super(app, plugin);
    this.plugin = plugin;
  }

  display(): void {
    let { containerEl } = this;

    containerEl.empty();

    containerEl.createEl("h2", { text: "Settings for my awesome plugin." });

    new Setting(containerEl)
      .setName("Setting #1")
      .setDesc("It's a secret")
      .addText((text) =>
        text
          .setPlaceholder("Enter your secret")
          .setValue("")
          .onChange(async (value) => {
            console.log("Secret: " + value);
            this.plugin.settings.mySetting = value;
            await this.plugin.saveSettings();
          })
      );
  }
}
