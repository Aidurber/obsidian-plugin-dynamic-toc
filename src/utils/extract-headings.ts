import { CachedMetadata, HeadingCache } from "obsidian";
import { Heading } from "../models/heading";
import { DynamicTOCSettings, TableOptions } from "../types";

function clean(heading: string): string {
  return heading.replace(/[[\]]/g, "");
}

function buildMarkdownText(headings: HeadingCache[], options: TableOptions) {
  const firstHeadingDepth = headings[0].level;
  return headings
    .map((heading) => {
      const itemIndication = (options.style === "number" && "1.") || "-";
      const indent = new Array(Math.max(0, heading.level - firstHeadingDepth))
        .fill("\t")
        .join("");
      const cleanHeading = clean(heading.heading);
      return `${indent}${itemIndication} [[#${cleanHeading}|${cleanHeading}]]`;
    })
    .join("\n");
}

function buildMarkdownTextV2(headings: Heading[], options: TableOptions) {
  const firstHeadingDepth = headings[0].level;

  return headings
    .map((heading) => {
      const itemIndication = (options.style === "number" && "1.") || "-";
      const indent = new Array(Math.max(0, heading.level - firstHeadingDepth))
        .fill("\t")
        .join("");

      return `${indent}${itemIndication} ${heading.markdownHref}`;
    })
    .join("\n");
}
export function extractHeadings(
  fileMetaData: CachedMetadata,
  options: TableOptions,
  settings: DynamicTOCSettings
) {
  if (!fileMetaData?.headings) return "";
  const { headings } = fileMetaData;
  const processableHeadings = headings.filter(
    (h) => !!h && h.level >= options.min_depth && h.level <= options.max_depth
  );

  if (!processableHeadings.length) return "";
  if (settings.useNewHeaderExtraction) {
    return buildMarkdownTextV2(
      processableHeadings.map((h) => new Heading(h)),
      options
    );
  }
  return buildMarkdownText(processableHeadings, options);
}
