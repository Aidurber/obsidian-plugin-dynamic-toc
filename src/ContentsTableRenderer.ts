import { Component, MarkdownRenderer, parseYaml, TFile } from "obsidian";
import { IHeadingExtractor } from "./HeadingExtractor";
import { DynamicTOCSettings, TableOptions } from "./types";

export class ContentsTableRenderer {
  constructor(
    protected headingExtractor: IHeadingExtractor,
    protected source: string,
    protected element: HTMLElement,
    protected component: Component,
    protected settings: DynamicTOCSettings
  ) {}

  private get options(): TableOptions {
    try {
      const options = parseYaml(this.source) as TableOptions;
      const merged = Object.assign({}, this.settings, options);
      return Object.keys(merged).reduce((acc, curr: keyof TableOptions) => {
        const value = options[curr];
        const isEmptyValue = typeof value === "undefined" || value === null;
        return {
          ...acc,
          [curr]: isEmptyValue ? this.settings[curr] : value,
        };
      }, {} as TableOptions);
    } catch (error) {
      return this.settings;
    }
  }

  /**
   * Build the table of contents
   * @param file - Current file
   */
  build = async (file: TFile) => {
    try {
      this.element.empty();
      this.element.classList.add("table-of-contents");
      const headings = this.headingExtractor.extract(file, this.options);
      await MarkdownRenderer.renderMarkdown(
        headings,
        this.element,
        file.path,
        this.component
      );
    } catch (error) {
      console.error(error);
    }
  };
}
