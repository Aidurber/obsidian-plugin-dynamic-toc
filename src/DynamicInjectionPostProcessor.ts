import { App, MarkdownPostProcessorContext, MarkdownRenderer } from "obsidian";
import { IHeadingExtractor } from "./HeadingExtractor";
import { DynamicTOCSettings } from "./types";
export class DynamicInjectionPostProcessor {
  constructor(
    protected headingExtractor: IHeadingExtractor,
    protected app: App,
    protected settings: DynamicTOCSettings
  ) {}
  process = async (
    el: HTMLElement,
    _: MarkdownPostProcessorContext
  ): Promise<any> => {
    if (!this.settings.injectionString?.length) return;
    let match: HTMLElement | null = null;
    try {
      const result = document.evaluate(
        `//*[text()[contains(.,'${this.settings.injectionString}')]]`,
        el
      );
      const firstResult = result.iterateNext();
      match = firstResult as HTMLElement;
    } catch (error) {
      console.error(error);
    }

    if (!match) return;
    const file = this.app.workspace.getActiveFile();
    const headings = this.headingExtractor.extract(file, this.settings);
    const newElement = document.createElement("div");
    newElement.classList.add("table-of-contents");

    await MarkdownRenderer.renderMarkdown(
      headings,
      newElement,
      file.path,
      undefined
    );
    match.parentNode.replaceChild(newElement, match);
  };
}
