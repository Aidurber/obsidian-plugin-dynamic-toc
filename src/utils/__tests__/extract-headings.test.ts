import { CachedMetadata } from "obsidian";
import { TableOptions } from "src/types";
import { extractHeadings } from "../extract-headings";

describe("Extract headings", () => {
  describe("build markdown text", () => {
    const defaultHeadings = {
      headings: [
        {
          heading: "foo",
          level: 1,
        },
        {
          heading: "bar",
          level: 2,
        },
        {
          heading: "baz",
          level: 3,
        },
        {
          heading: "[[Something|Alt Text]]",
          level: 4,
        },
        {
          heading: "level 1",
          level: 1,
        },
        {
          heading: "level 1 a",
          level: 2,
        },
      ],
    } as CachedMetadata;
    it("should match snapshot", () => {
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "number",
      } as TableOptions;
      expect(extractHeadings(defaultHeadings, options)).toMatchSnapshot();
    });

    it("should match snapshot when varied_style is true and style is bullet", () => {
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "bullet",
        varied_style: true,
      } as TableOptions;
      expect(extractHeadings(defaultHeadings, options)).toMatchSnapshot();
    });
    it("should match snapshot when varied_style is true and style is number", () => {
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "number",
        varied_style: true,
      } as TableOptions;
      expect(extractHeadings(defaultHeadings, options)).toMatchSnapshot();
    });

    it("should match snapshot with title", () => {
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "number",
        title: "## Table of Contents",
      } as TableOptions;
      expect(extractHeadings(defaultHeadings, options)).toMatchSnapshot();
    });

    it("should match snapshot with inconsistent heading levels", () => {
      const fileMetaData = {
        headings: [
          {
            heading: "Level 2",
            level: 2,
          },
          {
            heading: "Level 4",
            level: 4,
          },
          {
            heading: "Level 5",
            level: 5,
          },
          {
            heading: "Level 2",
            level: 2,
          },
          {
            heading: "Level 3",
            level: 3,
          },
        ],
      } as CachedMetadata;
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "number",
        allow_inconsistent_headings: true,
      } as TableOptions;
      expect(extractHeadings(fileMetaData, options)).toMatchSnapshot();
    });
  });
  describe("build inline markdown text", () => {
    const defaultHeadings = {
      headings: [
        {
          heading: "foo",
          level: 2,
        },
        {
          heading: "bar",
          level: 3,
        },
        {
          heading: "baz",
          level: 2,
        },
        {
          heading: "[[Something|Alt Text]]",
          level: 2,
        },
      ],
    } as CachedMetadata;
    it("should render correct markdown", () => {
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "inline",
      } as TableOptions;
      const result = extractHeadings(defaultHeadings, options);
      expect(result).toEqual(
        "[[#foo]] | [[#baz]] | [[#Something Alt Text|Alt Text]]"
      );
    });
    it("should accept a different delimiter", () => {
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "inline",
        delimiter: "*",
      } as TableOptions;
      const result = extractHeadings(defaultHeadings, options);
      expect(result).toEqual(
        "[[#foo]] * [[#baz]] * [[#Something Alt Text|Alt Text]]"
      );
    });
    it("should trim user delimiter", () => {
      const options = {
        max_depth: 4,
        min_depth: 1,
        style: "inline",
        delimiter: " * ",
      } as TableOptions;
      const result = extractHeadings(defaultHeadings, options);
      expect(result).toEqual(
        "[[#foo]] * [[#baz]] * [[#Something Alt Text|Alt Text]]"
      );
    });
  });
});
