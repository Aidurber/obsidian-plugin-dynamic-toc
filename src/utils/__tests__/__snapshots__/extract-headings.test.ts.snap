// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extract headings build markdown text should match snapshot 1`] = `
"1. [[#foo]]
	1. [[#bar]]
		1. [[#baz]]
			1. [[#Something Alt Text|Alt Text]]
1. [[#level 1]]
	1. [[#level 1 a]]"
`;

exports[`Extract headings build markdown text should match snapshot when varied_style is true and style is bullet 1`] = `
"- [[#foo]]
	1. [[#bar]]
		1. [[#baz]]
			1. [[#Something Alt Text|Alt Text]]
- [[#level 1]]
	1. [[#level 1 a]]"
`;

exports[`Extract headings build markdown text should match snapshot when varied_style is true and style is number 1`] = `
"1. [[#foo]]
	- [[#bar]]
		- [[#baz]]
			- [[#Something Alt Text|Alt Text]]
1. [[#level 1]]
	- [[#level 1 a]]"
`;

exports[`Extract headings build markdown text should match snapshot with inconsistent heading levels 1`] = `
"1. [[#Level 2]]
	1. [[#Level 4]]
1. [[#Level 2]]
	1. [[#Level 3]]"
`;

exports[`Extract headings build markdown text should match snapshot with title 1`] = `
"## Table of Contents
1. [[#foo]]
	1. [[#bar]]
		1. [[#baz]]
			1. [[#Something Alt Text|Alt Text]]
1. [[#level 1]]
	1. [[#level 1 a]]"
`;
